hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestStoreFileWriter.java

```diff
@@ -0,0 +1,353 @@
+    // allocating a separate column for them) in the memory representation of the data stored
+    // to HBase
+    for (int q = 0; q < qualifiers.length; q++) {
+      insertedCells.get(row).get(q).add(new CellInfo(timestamp, Cell.Type.DeleteFamilyVersion));
+    }
+  }
+
+  private void deleteRowVersions(int rowCount) throws Exception {
+    int row;
+    for (int r = 0; r < rowCount; r++) {
+      row = RANDOM.nextInt(ROW_NUM);
+      Long timestamp = getRowTimestamp(row);
+      if (timestamp != null) {
+        deleteSingleRowVersion(row, timestamp);
+      }
+    }
+    // Just insert one more delete marker possibly does not delete any row version
+    row = RANDOM.nextInt(ROW_NUM);
+    deleteSingleRowVersion(row, System.currentTimeMillis());
+  }
+
+  private void deleteColumns(int rowCount) throws Exception {
+    int row;
+    for (int r = 0; r < rowCount; r++) {
+      long timestamp = System.currentTimeMillis();
+      row = RANDOM.nextInt(ROW_NUM);
+      int q = RANDOM.nextInt(qualifiers.length);
+      Delete delete = new Delete(Bytes.toBytes(String.valueOf(row)), timestamp);
+      delete.addColumns(HBaseTestingUtil.fam1, qualifiers[q], timestamp);
+      regions[0].delete(delete);
+      regions[1].delete(delete);
+      insertedCells.get(row).get(q).add(new CellInfo(timestamp, Cell.Type.DeleteColumn));
+    }
+  }
+
+  private void deleteColumnVersions(int rowCount) throws Exception {
+    int row;
+    for (int r = 0; r < rowCount; r++) {
+      row = RANDOM.nextInt(ROW_NUM);
+      Long timestamp = getRowTimestamp(row);
+      if (timestamp != null) {
+        Delete delete = new Delete(Bytes.toBytes(String.valueOf(row)));
+        int q = RANDOM.nextInt(qualifiers.length);
+        delete.addColumn(HBaseTestingUtil.fam1, qualifiers[q], timestamp);
+        regions[0].delete(delete);
+        regions[1].delete(delete);
+        insertedCells.get(row).get(q).add(new CellInfo(timestamp, Cell.Type.Delete));
+      }
+    }
+  }
+
+  private Scan createScan(boolean raw) {
+    Scan scan = new Scan();
+    scan.readAllVersions();
+    scan.setRaw(raw);
+    return scan;
+  }
+
+  private void scanAndCompare(boolean raw) throws Exception {
+    try (RegionScanner firstRS = regions[0].getScanner(createScan(raw))) {
+      try (RegionScanner secondRS = regions[1].getScanner(createScan(raw))) {
+        boolean firstHasMore;
+        boolean secondHasMore;
+        do {
+          List<Cell> firstRowList = new ArrayList<>();
+          List<Cell> secondRowList = new ArrayList<>();
+          firstHasMore = firstRS.nextRaw(firstRowList);
+          secondHasMore = secondRS.nextRaw(secondRowList);
+          assertEquals(firstRowList.size(), secondRowList.size());
+          int size = firstRowList.size();
+          for (int i = 0; i < size; i++) {
+            Cell firstCell = firstRowList.get(i);
+            Cell secondCell = secondRowList.get(i);
+            assert (CellUtil.matchingRowColumn(firstCell, secondCell));
```

Comment:
nit: These asserts could be assertTrue or assertEquals, but this is not blocker for merging the PR.

Reply:
I made the change for this comment.